orchagent/twamporch.cpp

```diff
@@ -0,0 +1,1083 @@
+
+#define TWAMP_SESSION_TIMEOUT_MIN    1
+#define TWAMP_SESSION_TIMEOUT_MAX    10
+
+/* Default maximum number of TWAMP session */
+#define DEFAULT_NUMBER_OF_TWAMP_SESSION   0
+
+static map<string, sai_twamp_session_role_t> twamp_role_map =
+{
+    { "SENDER",       SAI_TWAMP_SESSION_ROLE_SENDER    },
+    { "REFLECTOR",    SAI_TWAMP_SESSION_ROLE_REFLECTOR }
+};
+
+static map<string, sai_twamp_mode_t> twamp_mode_map =
+{
+    { "FULL",     SAI_TWAMP_MODE_FULL    },
+    { "LIGHT",    SAI_TWAMP_MODE_LIGHT   }
+};
+
+static map<string, sai_twamp_timestamp_format_t> timestamp_format_map =
+{
+    { "NTP",    SAI_TWAMP_TIMESTAMP_FORMAT_NTP },
+    { "PTP",    SAI_TWAMP_TIMESTAMP_FORMAT_PTP }
+};
+
+static map<string, bool> session_admin_state_map =
+{
+    { "ENABLED",  true  },
+    { "DISABLED", false }
+};
+
+static map<string, bool> hw_lookup_map =
+{
+    { "TRUE",  true  },
+    { "FALSE", false }
+};
+
+/* Global variables */
+extern sai_object_id_t     gSwitchId;
+extern sai_object_id_t     gVirtualRouterId;
+extern sai_switch_api_t   *sai_switch_api;
+extern sai_twamp_api_t    *sai_twamp_api;
+extern CrmOrch            *gCrmOrch;
+
+const vector<sai_twamp_session_stat_t> twamp_session_stat_ids =
+{
+    SAI_TWAMP_SESSION_STAT_RX_PACKETS,
+    SAI_TWAMP_SESSION_STAT_RX_BYTE,
+    SAI_TWAMP_SESSION_STAT_TX_PACKETS,
+    SAI_TWAMP_SESSION_STAT_TX_BYTE,
+    SAI_TWAMP_SESSION_STAT_DROP_PACKETS,
+    SAI_TWAMP_SESSION_STAT_MAX_LATENCY,
+    SAI_TWAMP_SESSION_STAT_MIN_LATENCY,
+    SAI_TWAMP_SESSION_STAT_AVG_LATENCY,
+    SAI_TWAMP_SESSION_STAT_MAX_JITTER,
+    SAI_TWAMP_SESSION_STAT_MIN_JITTER,
+    SAI_TWAMP_SESSION_STAT_AVG_JITTER
+};
+
+
+
+TwampOrch::TwampOrch(TableConnector confDbConnector, TableConnector stateDbConnector, SwitchOrch *switchOrch, PortsOrch *portOrch, VRFOrch *vrfOrch) :
+        Orch(confDbConnector.first, confDbConnector.second),
+        m_stateDbTwampTable(stateDbConnector.first, stateDbConnector.second),
+        m_switchOrch(switchOrch),
+        m_portsOrch(portOrch),
+        m_vrfOrch(vrfOrch)
+{
+    /* Set entries count to 0 */
+    m_maxTwampSessionCount = m_twampSessionCount = 0;
+
+    /* Get the Maximum supported TWAMP sessions */
+    SWSS_LOG_INFO("Get the Maximum supported TWAMP sessions");
+    sai_attribute_t attr;
+    attr.id = SAI_SWITCH_ATTR_MAX_TWAMP_SESSION;
+    sai_status_t status = sai_switch_api->get_switch_attribute(gSwitchId, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_WARN("Failed to get switch attribute maximum TWAMP sessions.");
+        m_maxTwampSessionCount = DEFAULT_NUMBER_OF_TWAMP_SESSION;
+    }
+    else
+    {
+        m_maxTwampSessionCount = attr.value.s32;
+    }
+
+    /* Set MAX entries to counter DB */
+    vector<FieldValueTuple> fvTuple;
+    fvTuple.emplace_back("MAX_TWAMP_SESSION_COUNT", to_string(m_maxTwampSessionCount));
+    m_switchOrch->set_switch_capability(fvTuple);
+
+    /* Add TWAMP session event notification support */
```

Comment:
Why do we need to register this if SAI capability is returned as false? Please skip this section if SAI returned false for switch capability

Reply:
Done

Reply:
I didn't get the change here. My comment is to skip the entire registration process if SAI capability is returned as false. Where is the capability check?

Reply:
Need to check capability when creating entry, the default value 0 means no resources and return failed(see line 512, please).
Also I will add skip if the SAI capability is returned as false.
Thank you again for your comment.

Reply:
Done, all tests passed. Please review again, thank you very much.